cashweb-registry/src/store/topics.rs

```diff
@@ -82,25 +87,55 @@ impl<'a> DbTopics<'a> {
         if split_topic.iter().any(|segment| segment.is_empty()) {
             return Err(TopicInvalidSegments())?;
         }
-
         let payload_hash = message.payload_hash().as_slice().to_vec();
 
+        // Update existing record if we have one.
+        // TODO: needs a key-level lock on this.
+        let possibly_existing_message = self.get_message(&payload_hash);
+        let db = self.db.rocksdb();
+        let payload_buf = message.to_proto().encode_to_vec();
         let mut batch = rocksdb::WriteBatch::default();
 
-        batch.put_cf(self.cf_payloads, &payload_hash, &buf);
-
-        for idx in 0..split_topic.len() + 1 {
-            let base_topic_parts = split_topic[..idx].join(".");
-            let topic_digest = Sha256::digest(base_topic_parts.as_bytes().into())
-                .as_slice()
-                .to_vec();
-            let topical_key = [
-                &topic_digest,
-                timestamp.to_be_bytes().as_ref(),
+        let mut new_burns = vec![];
+        if possibly_existing_message.is_ok() {
+            let mut found_message = possibly_existing_message.unwrap();
+            new_burns.extend(found_message.add_burn_txs(message.txs()));
+            // Update existing key.
+            // TODO: Implement merging here instead of doing an update in this way as there may be a database race.
+            batch.put_cf(
+                self.cf_payloads,
                 &payload_hash,
-            ]
-            .concat();
-            batch.put_cf(self.cf_messages, &topical_key, &payload_hash);
+                &message.to_proto().encode_to_vec(),
+            );
+        } else {
+            new_burns.extend(message.txs());
+            //
+            batch.put_cf(self.cf_payloads, &payload_hash, &payload_buf);
+        }
+        // TODO: We need to only update the topics w/ timestamp iff the burn txns are not yet in there.
+        let buf = message.to_proto_without_payload().encode_to_vec();
+        for burn_tx in new_burns {
+            let tx_id = lotus_txid(burn_tx.tx().unhashed_tx()).to_vec_be();
```

Comment:
You have to be careful with the endianness here. Sometimes you're using `to_vec_be()` (big endian), sometimes you're using `as_slice()` (little endian). It's not clear to me if this code is correct.

I usually completely avoid big endian where possible (except for SLP, where I always use big endian...).

Reply:
Only one that matters is the timestamp. The others are just simple chunks.

Reply:
I think I made it more consistent w/ less conversions now.